crypto/x509/x509_vfy.c

```diff
@@ -3228,71 +3173,47 @@ static int build_chain(X509_STORE_CTX *ctx)
         }
 
         /*
-         * Extend chain with peer-provided certificates
+         * Extend chain with peer-provided untrusted certificates
          */
         if ((search & S_DOUNTRUSTED) != 0) {
             num = sk_X509_num(ctx->chain);
-            if (!ossl_assert(num == ctx->num_untrusted)) {
-                ERR_raise(ERR_LIB_X509, ERR_R_INTERNAL_ERROR);
-                trust = X509_TRUST_REJECTED;
-                ctx->error = X509_V_ERR_UNSPECIFIED;
-                search = 0;
-                continue;
-            }
-            x = sk_X509_value(ctx->chain, num-1);
-
-            /*
-             * Once we run out of untrusted issuers, we stop looking for more
-             * and start looking only in the trust store if enabled.
-             */
-            xtmp = (self_signed || depth < num) ? NULL
-                                                : find_issuer(ctx, sktmp, x);
-            if (xtmp == NULL) {
+            if (!ossl_assert(num == ctx->num_untrusted))
+                goto int_err;
+            curr = sk_X509_value(ctx->chain, num - 1);
+            if ((self_signed = X509_self_signed(curr, 0)) < 0)
```

Comment:
The added X509_self_signed() call here seems like substantial change. Could you please elaborate why it was needed?

Reply:
The `self_signed` variable should refer to the contents of the `cert` variable,
and just before `cert` was updated, and so should be `self_signed`.
I believe it was a bug not to set it, and right, this is a semantic change.

Reply:
I just found that I had inserted as part of this PR a similar assignment to `self_signed` also in line 3043, for the very same reason.
I still believe that this makes much sense in both cases also because just thereafter an issuer for `curr` is looked up, which would be pointless for self-signed (and thus self-issued) certs.
Yet since this PR (at least meanwhile) is supposed to do just refactoring and likely it would be good to discuss this more thoroughly also with @vdukhovni, I've just removed both these semantic changes and will make them the (only) topic of a follow-up PR.

Reply:
OK